sycl/source/detail/usm/usm_impl.cpp

```diff
@@ -230,5 +234,60 @@ void *aligned_alloc(size_t Alignment, size_t Size, const queue &Q, alloc Kind) {
   return aligned_alloc(Alignment, Size, Q.get_device(), Q.get_context(), Kind);
 }
 
+// get_pointer_info
+std::tuple<alloc, device> get_pointer_info(const void *Ptr,
+                                           const context &Ctxt) {
+  if (Ctxt.is_host()) {
+    return std::make_tuple(alloc::host, Ctxt.get_devices()[0]);
+  }
+  std::shared_ptr<detail::context_impl> CtxImpl = detail::getSyclObjImpl(Ctxt);
+  std::shared_ptr<detail::usm::USMDispatcher> Dispatch =
+      CtxImpl->getUSMDispatch();
+  pi_context C = CtxImpl->getHandleRef();
+  cl_unified_shared_memory_type_intel AllocTy;
+  pi_device DeviceId;
+
+  // All these CL enums should be replicated in PI
+  detail::pi::checkPiResult(
+      Dispatch->getMemAllocInfo(C, Ptr, CL_MEM_ALLOC_TYPE_INTEL,
+                                sizeof(cl_mem_info_intel), &AllocTy, nullptr));
+  detail::pi::checkPiResult(
+      Dispatch->getMemAllocInfo(C, Ptr, CL_MEM_ALLOC_DEVICE_INTEL,
+                                sizeof(pi_device), &DeviceId, nullptr));
+
+  alloc ResultAlloc;
+  if (AllocTy == CL_MEM_TYPE_HOST_INTEL) {
+    ResultAlloc = alloc::host;
+  } else if (AllocTy == CL_MEM_TYPE_DEVICE_INTEL) {
+    ResultAlloc = alloc::device;
+  } else if (AllocTy == CL_MEM_TYPE_SHARED_INTEL) {
+    ResultAlloc = alloc::shared;
+  } else {
+    ResultAlloc = alloc::unknown;
+  }
+
+  // check device id
+  device Result;
+  if (DeviceId == nullptr) {
```

Comment:
```
if (DeviceId == cl_device_id {}) {
```
or perhaps just
```
if (DeviceId == {}) {
```
if it does compile...

Reply:
The cl API returns NULL for the device ID if it's a host allocation.

Reply:
OK, I checked and `if (DeviceId == {})` does no compile...